### ISSUE TITLE
bootstrap.py failed when installing runas

### ISSUE THREAD

**Issue by capouch** [atom/electron#2434]
I am moving development to a new VM, and after I cloned the repo, bootstrap.py fails when trying to install runas:

** snip **
npm verb stack Error: runas@2.0.0 install: `node-gyp rebuild`
npm verb stack Exit status 1
npm verb stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/lifecycle.js:213:16)
npm verb stack     at EventEmitter.emit (events.js:110:17)
npm verb stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/spawn.js:24:14)
npm verb stack     at ChildProcess.emit (events.js:110:17)
npm verb stack     at maybeClose (child_process.js:1015:16)
npm verb stack     at Process.ChildProcess._handle.onexit (child_process.js:1087:5)
npm verb pkgid runas@2.0.0
npm verb cwd /usr/src/electron
npm ERR! Linux 3.18.11
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "--verbose"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! code ELIFECYCLE

npm ERR! runas@2.0.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the runas@2.0.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the runas package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls runas
npm ERR! There is likely additional logging output above.
npm verb exit [ 1, true ]
npm verb unbuild node_modules/runas
npm info preuninstall runas@2.0.0
npm info uninstall runas@2.0.0
npm verb unbuild rmStuff runas@2.0.0 from /usr/src/electron/node_modules
npm info postuninstall runas@2.0.0
npm verb gentlyRm don't care about contents; nuking /usr/src/electron/node_modules/runas

npm ERR! Please include the following file with any support request:
npm ERR!     /usr/src/electron/npm-debug.log
None
Traceback (most recent call last):
  File "./script/bootstrap.py", line 189, in <module>
    sys.exit(main())
  File "./script/bootstrap.py", line 41, in main
    update_node_modules('.')
  File "./script/bootstrap.py", line 127, in update_node_modules
    execute_stdout(args, env)
  File "/usr/src/electron/script/lib/util.py", line 176, in execute_stdout
    raise e
subprocess.CalledProcessError: Command '['npm', 'install', '--verbose']' returned non-zero exit status 1

**Comment by capouch** [comment]
Re-reading this it doesn't look like it was installing runas, but rather using it.

Sorry I don't have the chops to analyze further.

**Comment by zcbenz** [comment]
Does `npm install runas` work for you?

**Comment by capouch** [comment]
Yes, and I tried it both as a local user and with -g as root, both worked perfectly.

**Comment by zcbenz** [comment]
Can you paste the whole `npm-debug.log` in gist?

**Comment by capouch** [comment]
It's thousands of lines; I set my scroll buffer to 5000 and I'm still only picking up a small part of it.  Is there any way to compress the file and attach it?

**Comment by capouch** [comment]
https://gist.github.com/cbcc789295801e7b1f04.git

That was the best I could do.  It was dizzying.

**Comment by zcbenz** [comment]
That URL is 404 not found, maybe you can find a repo and upload it to GitHub releases.

**Comment by capouch** [comment]
I don't know how I was supposed to do it, but you can clone that gist with git clone and see most of the file.

**Comment by zcbenz** [comment]
I can not find out what's wrong in npm-delog.log :disappointed:, if you are simply building electron, you can ignore the error and simply continue building:

```bash
$ ./script/update.py
$ ./script/build.py
```

Maybe the building step can reveal the root problem.

**Comment by capouch** [comment]
brianc@basebox:~/electron$ script/build.py -c D
ninja: Entering directory `out/D'
[3/1105] CC obj/vendor/node/deps/openssl/openssl/ssl/openssl.d1_clnt.o
FAILED: /home/brianc/electron/vendor/llvm-build/Release+Asserts/bin/clang -MMD -MF obj/vendor/node/deps/openssl/openssl/ssl/openssl.d1_clnt.o.d -DL_ENDIAN -DPURIFY -D_REENTRANT -DOPENSSL_NO_SSL2 -DOPENSSL_NO_SSL3 -DOPENSSL_NO_HEARTBEATS '-DENGINESDIR="/dev/null"' -DTERMIOS '-DOPENSSLDIR="/etc/ssl"' -DOPENSSL_NO_ASM -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DV8_USE_EXTERNAL_STARTUP_DATA -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_BITMAP_CONFIG -DSK_SUPPORT_LEGACY_DEVICE_VIRTUAL_ISOPAQUE -DSK_SUPPORT_LEGACY_N32_NAME -DSK_SUPPORT_LEGACY_SETCONFIG -DSK_IGNORE_ETC1_SUPPORT -DSK_IGNORE_GPU_DITHER -DTOOLKIT_VIEWS -DUSE_AURA -DUSE_X11 -DUSE_NSS_CERTS -DUSE_NSS -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUG -DCOMPONENT_BUILD -DGURL_DLL -DSKIA_DLL -DUSING_V8_SHARED -DWEBKIT_DLL -I../../vendor/node/deps/openssl -I../../vendor/node/deps/openssl/openssl -I../../vendor/node/deps/openssl/openssl/crypto -I../../vendor/node/deps/openssl/openssl/crypto/asn1 -I../../vendor/node/deps/openssl/openssl/crypto/evp -I../../vendor/node/deps/openssl/openssl/crypto/md2 -I../../vendor/node/deps/openssl/openssl/crypto/modes -I../../vendor/node/deps/openssl/openssl/crypto/store -I../../vendor/node/deps/openssl/openssl/include -Wno-missing-field-initializers -Wno-parentheses-equality -Wno-unused-function -Wno-sometimes-uninitialized -Wno-pointer-sign -Wno-string-plus-int -Wno-unused-variable -Wno-unused-value -Wno-deprecated-declarations -Wno-return-type -fPIC -fvisibility=hidden -m64 -march=x86-64   -c ../../vendor/node/deps/openssl/openssl/ssl/d1_clnt.c -o obj/vendor/node/deps/openssl/openssl/ssl/openssl.d1_clnt.o
/home/brianc/electron/vendor/llvm-build/Release+Asserts/bin/clang: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory
[3/1105] ACTION compile_coffee: compil...offee_e42ad8b794496e05f38fce7e796ecf89
ninja: build stopped: subcommand failed.
brianc@basebox:~/electron$ ./script/update.py 
brianc@basebox:~/electron$ ./script/build.py 
ninja: Entering directory `out/R'
[2/1019] CC obj/vendor/node/deps/zlib/zlib.inftrees.o
FAILED: ../../vendor/llvm-build/Release+Asserts/bin/clang -MMD -MF obj/vendor/node/deps/zlib/zlib.inftrees.o.d -DZ_HAVE_UNISTD_H -DNDEBUG -DV8_USE_EXTERNAL_STARTUP_DATA -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_BITMAP_CONFIG -DSK_SUPPORT_LEGACY_DEVICE_VIRTUAL_ISOPAQUE -DSK_SUPPORT_LEGACY_N32_NAME -DSK_SUPPORT_LEGACY_SETCONFIG -DSK_IGNORE_ETC1_SUPPORT -DSK_IGNORE_GPU_DITHER -DTOOLKIT_VIEWS -DUSE_AURA -DUSE_X11 -DUSE_NSS_CERTS -DUSE_NSS -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -I../../vendor/node/deps/zlib -Wno-parentheses-equality -Wno-unused-function -Wno-sometimes-uninitialized -Wno-pointer-sign -Wno-string-plus-int -Wno-unused-variable -Wno-unused-value -Wno-deprecated-declarations -Wno-return-type -fPIC -m64 -march=x86-64 -O2 -g -fno-ident -fdata-sections -ffunction-sections   -c ../../vendor/node/deps/zlib/inftrees.c -o obj/vendor/node/deps/zlib/zlib.inftrees.o
../../vendor/llvm-build/Release+Asserts/bin/clang: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory
[2/1019] CC obj/vendor/node/deps/zlib/zlib.trees.o
FAILED: ../../vendor/llvm-build/Release+Asserts/bin/clang -MMD -MF obj/vendor/node/deps/zlib/zlib.trees.o.d -DZ_HAVE_UNISTD_H -DNDEBUG -DV8_USE_EXTERNAL_STARTUP_DATA -DSK_SUPPORT_LEGACY_GETTOPDEVICE -DSK_SUPPORT_LEGACY_BITMAP_CONFIG -DSK_SUPPORT_LEGACY_DEVICE_VIRTUAL_ISOPAQUE -DSK_SUPPORT_LEGACY_N32_NAME -DSK_SUPPORT_LEGACY_SETCONFIG -DSK_IGNORE_ETC1_SUPPORT -DSK_IGNORE_GPU_DITHER -DTOOLKIT_VIEWS -DUSE_AURA -DUSE_X11 -DUSE_NSS_CERTS -DUSE_NSS -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -I../../vendor/node/deps/zlib -Wno-parentheses-equality -Wno-unused-function -Wno-sometimes-uninitialized -Wno-pointer-sign -Wno-string-plus-int -Wno-unused-variable -Wno-unused-value -Wno-deprecated-declarations -Wno-return-type -fPIC -m64 -march=x86-64 -O2 -g -fno-ident -fdata-sections -ffunction-sections   -c ../../vendor/node/deps/zlib/trees.c -o obj/vendor/node/deps/zlib/zlib.trees.o
../../vendor/llvm-build/Release+Asserts/bin/clang: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory
ninja: build stopped: subcommand failed.

**Comment by capouch** [comment]
I am still successfully building on the previous VM.  The only difference between the two is that I did not install clang on the new one, as it looked like there was a built-in version?

**Comment by zcbenz** [comment]
You can follow the troubleshooting to fix this:
https://github.com/atom/electron/blob/master/docs/development/build-instructions-linux.md

Not long ago Electron started to use pre-built clang binaries provided by Chromium for building, so we can reduce the requirements of building machine to minimum.

**Comment by capouch** [comment]
I got electron to build, and I believe there is a bug in the built-in clang implementation.  Some time ago libncurses was split into two pieces, libncurses and libtiinfo.  (see https://forums.opensuse.org/showthread.php/446927-missing-library-libtinfo-so-5).

The built-in clang++ compiler requires a number of system libraries, and in my case, I have the older (no libtinfo.so, only ncurses.so) implementation.  Because there is no libtinfo.so included with the pre-built clang binary, and it also was not in my system libraries, the entire build failed.

The fix was for me to place a symlink in ~/electron/vendor/llvm-build/Release+Asserts/lib to my system libncurses.so library.  Once that was done, the build proceeded perfectly.  Others without libtinfo.so, I believe, will experience the same build failure that I did.
